pallets/gear-builtin/src/bls12_381.rs

```diff
@@ -143,29 +144,46 @@ fn multi_miller_loop<T: Config>(
         Ok(count_b) if count_b != count => {
             return (
                 Err(BuiltinActorError::Custom(LimitedStr::from_small_str(
-                    "Multi Miller loop: non equal item count",
+                    "Multi Miller loop: uneven item count",
                 ))),
-                gas_spent,
+                gas_limit.saturating_sub(gas_left),
+            )
+        }
+        Err(_) => {
+            return (
+                Err(BuiltinActorError::DecodingError),
+                gas_limit.saturating_sub(gas_left),
             )
         }
-        Err(_) => return (Err(BuiltinActorError::DecodingError), gas_spent),
         Ok(_) => (),
     }
 
     let to_spend = <T as Config>::WeightInfo::bls12_381_multi_miller_loop(count as u32).ref_time();
-    if gas_limit < gas_spent + to_spend {
-        return (Err(BuiltinActorError::InsufficientGas), gas_spent);
+    if gas_left < to_spend {
+        return (
+            Err(BuiltinActorError::InsufficientGas),
+            gas_limit.saturating_sub(gas_left),
+        );
+    }
+    if GasAllowanceOf::<T>::get() < to_spend {
+        return (
+            Err(BuiltinActorError::GasAllowanceExceeded),
+            gas_limit.saturating_sub(gas_left),
+        );
     }
 
-    gas_spent += to_spend;
+    gas_left -= to_spend;
```

Comment:
saturating?

Reply:
Safe due to a check above - if `gas_left` is less than `to_spend` we abort earlier

Reply:
But for the sake of universality we can use saturated arithmetics - no problem